src/Microsoft.OData.Core/ObjectPool.cs

```diff
@@ -0,0 +1,100 @@
+
+using System;
+
+namespace Microsoft.OData
+{
+    /// <summary>
+    /// Manages a pool of reusable objects. An object retrieved from the pool should be returned when it's no longer being used so that it can be later recycled.
+    /// </summary>
+    /// <typeparam name="T">The type of objects to pool.</typeparam>
+    /// <remarks>This object pool is not thread-safe and not intended to be shared across concurrent requests.
+    /// Consider using the DefaultObjectPool in 8.x once we drop support for netstandard 1.1
+    /// </remarks>
+    internal class ObjectPool<T> where T : class
+    {
+        private readonly Func<T> objectGenerator;
+        private ObjectWrapper[] items;
+        private T firstItem;
+
+        // The pool size is equal to the level of nesting in the response.
+        // 32 is an arbitrary figure. Could be adjusted appropriately.
+        private const int POOL_SIZE = 8;
+
+        /// <summary>
+        /// To initialize the object pool.
+        /// </summary>
+        /// <param name="objectGenerator">Used to create an instance of a <typeparamref name="T"/>.</param>
+        public ObjectPool(Func<T> objectGenerator)
+        {
+            this.objectGenerator = objectGenerator ?? throw new ArgumentNullException(nameof(objectGenerator));
+        }
+
+        /// <summary>
+        /// Gets an object from the pool if one is available, otherwise creates one. The object will need to be returned to the pool when no longer in use.
+        /// </summary>
+        /// <returns>An object of type <typeparamref name="T"/> from the pool.</returns>
+        /// <remarks>It's the responsibility of the caller to clean up the object before using it.</remarks>
+        public T Get()
+        {
+            T item = firstItem;
+
+            if (item != null)
+            {
+                firstItem = default(T);
+                return item;
+            }
+
+            if (items == null)
+            {
+                items = new ObjectWrapper[POOL_SIZE];
+            }
+
+            for (int i = 0; i < items.Length; i++)
+            {
+                item = items[i].Element;
+
+                if (item != null)
+                {
+                    items[i].Element = default(T);
+                    return item;
+                }
+            }
+
+            return objectGenerator();
+        }
+
+        /// <summary>
+        /// Return an object to the pool.
+        /// </summary>
+        /// <param name="obj">The object to return to the pool.</param> 
+        public void Return(T obj)
+        {
+            if (firstItem == null)
+            {
+                firstItem = obj;
+                return;
+            }
+
+            for (int i = 0; i < items.Length; ++i)
+            {
+                if (items[i].Element == null)
+                {
+                    items[i].Element = obj;
+                    break;
+                }
+            }
+        }

```

Comment:
It's strange that if we get to the end without being able to return the object, nothing happens. The pool is fixed size, but there is nothing to the caller that indicates this, and if they *do* hit the limit, they will never notice because then we will just start generated new instances that leave scope and get garbage collected, removing the whole point of using the pool.

Reply:
I think when we hit the limit we have a number of options:
- throw an exception
- grow the pool
- do nothing

- I personally don't thin throwing an exception is a good idea. The object pool is just an optimization, an implementation detail. It shouldn't impose a limit on how much nesting the writer should support (the nesting limit is defined in the writer settings I believe, and the default is much higher, maybe 100).
- Doing nothing means the array does not need to be resized. It means for sufficiently shallow responses (we assume that's the majority of responses) the fixed size is good enough. And for more deeply nested responses, the performance would not be worse than the current implementation as far as allocations are concerned.
- Growing the pool means allocating a new array and copying over items, like `Array.Resize`. Maybe this in acceptable compromise for deeply-nested responses since the resizing is infrequent. What do you think about this? And if we allow resizing, would it make sense to start with a smaller array size (say 4 instead of 8)?